python/ray/data/_internal/iter_batches.py

```diff
@@ -0,0 +1,934 @@
+import collections
+import heapq
+import itertools
+import logging
+import random
+import queue
+import threading
+import sys
+from typing import (
+    Any,
+    Callable,
+    Iterator,
+    List,
+    NamedTuple,
+    Optional,
+    Tuple,
+    TypeVar,
+    Union,
+)
+
+import ray
+from ray.actor import ActorHandle
+from ray.data._internal.batcher import Batcher, ShufflingBatcher, _blocks_to_batches
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data._internal.stats import DatasetPipelineStats, DatasetStats
+from ray.data._internal.memory_tracing import trace_deallocation
+from ray.data.block import Block, BlockAccessor, BlockMetadata, DataBatch
+from ray.data.context import DatasetContext
+from ray.types import ObjectRef
+from ray.util.scheduling_strategies import NodeAffinitySchedulingStrategy
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+logger = logging.getLogger(__name__)
+
+PREFETCHER_ACTOR_NAMESPACE = "ray.dataset"
+
+if sys.version_info >= (3, 7):
+    from contextlib import nullcontext
+else:
+    from contextlib import contextmanager
+
+    @contextmanager
+    def nullcontext(enter_result=None):
+        yield enter_result
+
+
+def iter_batches(
```

Comment:
This function sounds very innocuous when it's doing a lot under the hood (creating threads, etc.).

How about making it a class, such as `ThreadPoolBatcher`, to make it more clear when reading code that we're setting up this state and executing batching in parallel?